src/librustc/traits/error_reporting.rs

```diff
@@ -117,8 +118,22 @@ impl<'a, 'gcx, 'tcx> Visitor<'gcx> for FindLocalByTypeVisitor<'a, 'gcx, 'tcx> {
 
 impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     pub fn report_fulfillment_errors(&self, errors: &Vec<FulfillmentError<'tcx>>) {
+        let mut fn_errors = FxHashSet();
         for error in errors {
-            self.report_fulfillment_error(error);
+            // Avoid multiple errors saying the same thing for closures
```

Comment:
You can see the effect of this in [the stderr output](https://github.com/rust-lang/rust/pull/41488/commits/d3f64a636e0fd6d72a533ada73f5bb2716048c9e#diff-efb255e33f30b3be4691c53a0d69e155), right now for closure type mismatch errors we output two (may be three in some cases?, haven't seen that yet) after the compiler tries to coerce to `Fn`, `FnMut` and `FnOnce`. This change was hacky and want to fix it properly somewhere closer to the source but this is as close as I got to the culprit last night. We can either remove this hack now and merge the PR with the duplicated errors, or merge this as is and I send a follow up PR fully addressing this.